### PR TITLE
Add speed conversions and fix Mmo

### PR DESCRIPTION
**Pull request by MisterChocker**
- Fixed Mmo conversion was wrong
- Added Vmo 350kt and Mmo .82 as constant
- Added temperature conversion °C to K
- Added speed conversions:
   - TAS to KCAS
   - KCAS to KTAS
   - KTAS to MACH
   - MACH to KTAS
   - MACH to TCAS

## Screenshots (if necessary)

Before:
image

After:
image

## References

images

## Additional context

Discord username (if different from GitHub): DerL30N#3751

## Testing instructions

Test if Vmax is Min(Vfe, Vmo, Mmo)
Vfe: Max Flaps Speed
Vmo: 350kt
Mmo: .82M

Check overspeed warning is correct, check Vmax during climb or descent is either Vmo or Mmo whichever is smaller is displayed.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Review by Benjozork — APPROVED** [review]
LGTM

File: A32NX/html_ui/Pages/A32NX_Utils/NXSpeeds.js [hunk 1]
```diff
@@ -376,6 +379,77 @@ function _getVfeNIdx(fi) {
     }
 }
 
+/**
+ * Convert degrees Celsius into degrees Kelvin
```
Comment (tracernz) [inline A32NX/html_ui/Pages/A32NX_Utils/NXSpeeds.js:383]: No such thing as degrees Kelvin.. it's just Kelvin.

**Comment by donstim** [comment]
QA testers be aware, this is a bit of a hybrid, which can make it a bit tricky to test. The backstory is this:  The conversion Asobo uses from Mach to KCAS is wrong, and this PR adds the correct speed conversions, **but** does not display them yet with one exception. That exception is the KCAS that the M 0.82 MMO will now (under this PR) be displayed at on the PFD speed tape.  This means that the bottom of the upper red and black speed band, which corresponds to MMO, will be displayed at a KCAS speed on the speed tape that should be correct, but that KCAS will not be at what MSFS currently thinks is 0.82 Mach.

Here is an example:  Altitude 29,000 feet, ISA temp. MMO is shown at 319 KCAS, but when flying at 319 KCAS, MSFS thinks this is M0.80, well below MMO of M0.82. 
Screenshot (515)

To verify that MMO of 0.82 **is** 319 KCAS for this condition, consult an airspeed conversion calculator like this one:  https://aerotoolbox.com/airspeed-conversions/. For the above condition, one can verify that 319 KCAS is indeed M0.82:
Screenshot (517)

For my tests, I checked:
1. That the PFD speed tape VMAX was VMO with flaps up when MMO is greater than 350 knots. **Pass**
2. That VMAX was MMO when MMO is less than 350 knots. **Pass**
3. That VMAX was VFE for the associated flap setting when flaps are deployed (checked for each flap position). **Pass**
4. That an aural and ECAM overspeed warning occurred at the correct speed (more on that below)

The aural overspeed warning and ECAM message occur at speeds slightly above VMO/MMO as follows:
Screenshot (518)
Screenshot (520)

The VMO warning can be checked fairly easily, and it seemed to be right on. The MMO is a bit more difficult to check due to the speed conversion issue discussed above and because MSFS wants to switch to Mach mode automatically, which makes it difficult to fly a target KCAS. Nevertheless, the warning seemed to occur within a knot of the expected speed (maybe a knot high if I wanted to quibble.)

Not part of this PR, but is the Overspeed warning correctly displayed on the ECAM?  According to the FCOM material in the screenshot above, the ECAM warning should be "O/SPEED" rather than the "OVERSPEED" it displays now.

**Comment by beheh** [comment]
> Not part of this PR, but is the Overspeed warning correctly displayed on the ECAM? According to the FCOM material in the screenshot above, the ECAM warning should be "O/SPEED" rather than the "OVERSPEED" it displays now.

I have some WIP on my main machine for the updated overspeed warnings also for the correct flap speeds. I'll submit a PR after the holidays. No changes to the text though (outsideof maybe `OVERSPEED` => `OVER SPEED`):

image

**Comment by donstim** [comment]
> > Not part of this PR, but is the Overspeed warning correctly displayed on the ECAM? According to the FCOM material in the screenshot above, the ECAM warning should be "O/SPEED" rather than the "OVERSPEED" it displays now.
> 
> I have some WIP on my main machine for the updated overspeed warnings also for the correct flap speeds. I'll submit a PR after the holidays. No changes to the text though (outsideof maybe `OVERSPEED` => `OVER SPEED`):
> 
Okay. I was just a bit confused over the FCOM reference to "O/SPEED" warning instead of "OVERSPEED" warning. It must have just been an abbreviation used in that one instance.

I've never seen it shown as two words.  Everywhere I've seen it it's always been one word on the A320 ECAM. Also, in our mod, if VMAX is a VFE speed, does our overspeed warning also reference the associated VFE speed, as in -
Screenshot (524)

Will our overspeed warning be triggered for a VLE overspeed?

**Comment by MisterChocker** [comment]
Yes you get overspeed for when exceeding flaps speeds as well.

**Comment by donstim** [comment]
> Yes you get overspeed for when exceeding flaps speeds as well.

Yes, and I checked that as you can see in my write-up.

> Not sure if it currently says -FVE .... 185, as in your example.

It doesn't, but I'm sure it will after @beheh gets through with it!

**Comment by Lazarus320** [comment]
Category: QA Tester
Name: Lazarus#4286
Date of testing: /2021
Version of the sim: 1.12.13.0
PR Tested: #2781 
Object of test: Speed conversions and fix Mmo
Testing technique: Tested vmmo, mmo and aural overspeed warnings 
Steps you've made to test the Pull Request:

1. MMO 0.82 tested at 39000 feet 
image
   
2. VMO 350kts tested 
image

3. Aural warning at 350+4kts checked 

https://user-images.githubusercontent.com/74972083/103635556-aa376800-4f83-11eb-9d1a-78693f566022.mp4




Flight notes (If required): N/A

Issues (If Any): N/A
Overall Rating: Good
Conclusions of the testing: Looks good nice work.

**Comment by donstim** [comment]
When this gets merged, may want to alert the moderators of the Discord help forum. The way the autopilot seems to like to overshoot the Mach speed, I expect there will be a lot more overspeed cases when the overspeed warning barber pole is lowered to the correct KCAS value. We may see a lot of questions about that on the help channel.

**Comment by MisterChocker** [comment]
> When this gets merged, may want to alert the moderators of the Discord help forum. The way the autopilot seems to like to overshoot the Mach speed, I expect there will be a lot more overspeed cases when the overspeed warning barber pole is lowered to the correct KCAS value. We may see a lot of questions about that on the help channel.

when this get's merged I will add a feature that will prevent the athr from over speeding. If it works the way I thing it does, it's very simple to be implement.

**Comment by FlightSimmer7700** [comment]
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 6/1/21
Version of the sim1.12.13.0
PR Tested: #2781 
Object of test: Speed

Description of how you tested the Pull Request: Tested at different speeds and altitudes
Steps you've made to test the Pull Request:

Testing technique: 1. Loaded in at the runway
2. Climed up
3. Checked at different speeds and altitudes

Flight notes (If required):

Issues (If Any): 
Overall Rating: Good
Conclusions of the testing: All good, no issues found

**Comment by TwitchyShell** [comment]
**Category:** QA Tester
**Name:** Тайга Мөңгүн#2000
**Date of testing:** 2021-01-07
**Version of the sim:** 1.12.13.0
**PR Tested:** #2781 
**Object of test:** Testing VMO, MMO fix

# **Testing technique:** 
Full line flight, 1 sector initially, UHWW to UHSS, then UHHH to UHWW for verification. Initially using Meteoblue live weather.

## **Steps you've made to test the Pull Request:**
### Flight 1:
1. Preflight as normal
2. TO and departure
3. Climb in selected speed to induce overspeed
4. Normal cruise, calculate and compare speed readpits
5. Descent as normal
6. During approach induce overspeed, observe Vfe through the stages

### Flight 2:
1. Preflight, TO and dep as normal
2. Climb and cruise, observe KCAS vs Mach, and MMO readouts
3. Des, land, shutdown as per normal line flight

# Flight notes:
### **Flight 1**
Normal preflight and take off. Climbed and increased speed bug to past Vmax. Observed Overspeed warning on ECAM, Master Warning aural triggered as expected at VMO+4. (Figure 1.1)

Microsoft Flight Simulator Screenshot 2021 01 04 - 06 45 11 01 Thumbnail
**Figure 1.1**

Resumed normal climb, then cruise. Observed VMO MMO crossover at ~24500'-25500'. No figure captured. Let me know if a more accurate review of this is required.

Cruise and descent observed issues due to operator error, rectified in second attempt, otherwise nil comment.

During approach, verified speeds for Vfe before and after extension, observing Vmax adjusting as appropriate. Nil issues, behaviour as expected.

### Conf  1 - 230 KCAS
C1
**Figure 2.1**

### Conf 2 - 200 KCAS
C2
**Figure 2.2**

### Conf 3 -185 KCAS
C3
**Figure 2.3**

### Conf F - 177 KCAS
C4
**Figure 2.4**

Landing and shutdown nil comments, not of scope.

### **Flight 2**
Preflight, TO and dep nil comment.
Observed on climb proper VMO MMO crossover at suitable altitude as mentioned above.
Cruise, observed correct KCAS vs mach, as well as KCAS on MMO for appropriate conditions. Tested for ISA, ISA -17, ISA+11.5. All satisfactory. ISA +11.5 excerpt included below. MMO overspeed ECAM warning, aural and master warning observed as appropriate at MMO +.001, as mentioned by you, difficult to check as per reference parametres from manuals.

Microsoft Flight Simulator Screenshot 2021 01 07 - 07 20 14 27 Thumbnail
**Figure 3.1 - Simulator excerpt of flight at 38,000', ISA+11.5, OAT -45C**

image
**Figure 3.2 - Excerpt from airspeed conversion webpage to confirm MMO as per Figure 3.1**

image
**Figure 3.3 - Excerpt from airspeed conversion webpage to confirm current mach from TAS as per Figure 3.1**

Descent, observed VMO MMO crossover as mentioned above.
Approach, observed proper Vfe and Vmax limits.
Landing, shutdown nil comments, not of scope.

# **Issues (If Any):** Nil

### **Overall Rating:**
Good. Nil issues observed with this fix. Functions as described with behaviour within documented expectations.

Conclusions of the testing: Is work, very nice.

**Comment by MisterChocker** [comment]
@TwitchyShell thanks for testing!

**Pull request merged**